File: src/decoders/geolocated-instance.ts
import type { GeolocatedInstance } from "~/models";
import { haversine } from "~/core/haversine";

export const decodeGeolocatedInstance = (instance: any, position: [lat: number, lon: number]): GeolocatedInstance => {
  const latitude = parseFloat(instance.lat);
  const longitude = parseFloat(instance.long);

  return {
    url: instance.url.toLowerCase(),
    name: instance.nomEtab
      .trim()
      .replace("COLLEGE", "COLLÈGE")
      .replace("LYCEE", "LYCÉE"),

    latitude, longitude,
    postalCode: parseInt(instance.cp),

    distance: haversine(position, [latitude, longitude])
  };
};

File: library/src/commonMain/kotlin/decoders/GeolocatedInstance.kt
package decoders

import haversine
import kotlinx.serialization.json.*
import models.GeolocatedInstance

fun decodeGeolocatedInstance (instance: JsonObject, position: Pair<Double, Double>): GeolocatedInstance {
    val latitude = instance["lat"]!!.jsonPrimitive.double
    val longitude = instance["long"]!!.jsonPrimitive.double

    return GeolocatedInstance(
        url = instance["url"].toString().lowercase(),
        name = instance["nomEtab"].toString()
            .trim()
            .replace("COLLEGE", "COLLÈGE")
            .replace("LYCEE", "LYCÉE"),
        latitude = latitude,
        longitude = longitude,
        postalCode = instance["cp"]!!.jsonPrimitive.int.toUInt(),
        distance = haversine(position, Pair(latitude, longitude))
    )
}